### ISSUE TITLE
^0.5 is broken for LanguageClient-neovim, likely others

### ISSUE THREAD

**Issue by pappasam** [redhat-developer/yaml-language-server#192]
First, thanks for this language server. It's awesome and I look forward to it becoming even more awesome.

## Problem

A combination of changes between 0.4.1 and 0.5.0 broke compatibility with my LSP client (LanguageClient-neovim, latest version, commit hash ee6cdb0940d9c2c95f31488f7d463b2d12abcbc4. Using neovim 0.5.0 dev). I suspect it's broken for other Vim/neovim clients as well.

## Error with yaml-language-server version ^0.5

"Cannot read property \'hierarchicalDocumentSymbolSupport\' of undefined"

## Plethora of problems

Line 351 of src/server.ts contains the following line:

```javascript
hierarchicalDocumentSymbolSupport = !!capabilities.textDocument.documentSymbol.hierarchicalDocumentSymbolSupport;
```

This errors out when the server capabilities does not contain a documentSymbol. The LSP specification allows a client to omit this information, but the yaml-language-server does not elegantly handle this case (hierarchicalDocumentSymbolSupport is called on the undefined "documentSymbol" attribute).

I tried updating this line to:

```javascript
hierarchicalDocumentSymbolSupport = !!(capabilities.textDocument.documentSymbol && capabilities.textDocument.documentSymbol.hierarchicalDocumentSymbolSupport);
```

But now I receive a new error:

```
[LC] [Error] Notification handler 'workspace/didChangeConfiguration' failed with message: workspaceFolders is not iterable
```

Therefore, **the remedy is more complex than simply catching the "undefined" case.**

## Full original error message

```
WARN unnamed src/language_server_protocol.rs:2217 Failed to start language server automatically. Error: Failure { jsonrpc: Some(V2), error: Error { code: InternalError, message: "Request intialize failed with message: Cannot read property \'hierarchicalDocumentSymbolSupport\' of undefined", data: None }, id: Num(1) }
```

## Evidence from coc.nvim

Coc's yaml extension pins this library to 0.4.0, which strengthens my suspicion that yaml-language-server is broken for other vim language clients; coc users wouldn't notice the breakage because they don't use the latest version.

https://github.com/neoclide/coc-yaml/blob/master/package.json

```json
  "dependencies": {
    "yaml-language-server": "^0.4.0"
  }
```

## Reference

* [Specification section on Initialization](https://microsoft.github.io/language-server-protocol/specifications/specification-3-14/#initialize)

**Comment by fbricon** [comment]
@JPinkney you need to check if the client capabilities support that feature, during initialization

**Comment by bollwyvl** [comment]
Indeed! it looks like _all_ of the properties of `TextDocumentClientCapabilities` and `DocumentSymbolClientCapabilities` are optional:

https://github.com/microsoft/vscode-languageserver-node/blob/9edf211ab020c6bef23d7095014f3fe33c3af732/protocol/src/protocol.ts#L371

**Comment by bollwyvl** [comment]
Perhaps something as blunt as:
```ts
    workspaceFolders = params.workspaceFolders || [];

    try {
        hierarchicalDocumentSymbolSupport = !!capabilities.textDocument.documentSymbol.hierarchicalDocumentSymbolSupport;
    } catch { }

    try {
        clientDynamicRegisterSupport = !!(capabilities.textDocument.rangeFormatting && capabilities.textDocument.rangeFormatting.dynamicRegistration);
    } catch { }
```

**Comment by pappasam** [comment]
@JPinkney I confirm that @bollwyvl  's proposed "blunt" change fixes the issue, at least for me on LanguageClient-neovim. Not sure if that's the change we ultimately want to see in this codebase, but it's a potential path forward.

**Comment by JPinkney** [comment]
I'm still trying to get neovim working and everything installed but can we do something like:
```
if (!!capabilities.textDocument && !!capabilities.textDocument.documentSymbol......)
```
instead?

@bollwyvl do you mind submitting a pr with this code?

**Comment by bollwyvl** [comment]
I'd be happy to PR, but am not strictly a user or downstream... kind of
still surveying the language server space, and finding a fair number of
edge cases like this. As a spec, it seems rather loosely implemented across
the ecosystem. Perhaps the versioned headers (once broadly supported) will
help.

IIRC in this case there are multiple levels of potential undefinedness...
lodash.get or something, if you have that around...

I guess a !!((x || {}).y || {}).z would be reasonable as well.

That being said, I'd take the readability/performance trade of a
conservative default and try/catch for a once-per-connection, unversioned,
unvalidated, untrusted data, most days.

**Comment by JPinkney** [comment]
I've just released 0.6.0 which contains these fixes!
